policy_agent/policy_agent.py

```diff
@@ -0,0 +1,643 @@
+                _log.warning(
+                    "Apply %s to Pod %s : Profile Already exists" % (ns_tag, self.key))
+                pass
+            return True
+        else:
+            _log.error("Pod Resource %s found before Namespace Resource %s" % (
+                self.name, self.namespace))
+            return False
+
+    def remove_default_profile(self):
+        """
+        remove the default reject all rule programmed by the plugin
+        """
+        default_profile = "REJECT_ALL"
+        try:
+            _log.info("Removing Default Profile")
+            _datastore_client.remove_profiles_from_endpoint(profile_names=[default_profile],
+                                                            endpoint_id=self.ep_id)
+            return True
+        except ProfileNotInEndpoint:
+            _log.info("Default Profile not on Endpoint")
+            return True
+        else:
+            _log.info("Default Profile Removal Failed")
+            return False
+
+
+class Endpoints(Resource):
+
+    def from_json(self, json):
+        self.kind = "Endpoints"
+        self.name = json["metadata"]["name"]
+        self.namespace = json["metadata"]["namespace"]
+        self.subsets = json["subsets"]
+
+    def get_key(self):
+        return "%s/%s" % (self.namespace, self.name)
+
+    def generate_svc_profiles_pods(self):
+        """
+        Endpoints objects contain a list of subsets
+        containing pod and policy info
+        subsets: [
+            {
+                addresses: [
+                    ...pods...
+                ],
+                ports: [
+                    ...associated ports...
+                ]
+            },
+            {
+                addresses: [
+                    ...pods...
+                ],
+                ports: [
+                    ...associated ports...
+                ]
+            },
+        ]
+        :return: A generated dict of profile-pod associations
+        :rtype: a dict of profiles mapping to a list of assoc pods
+        """
+        def verify_args(ports_obj, try_arg):
+            try:
+                arg = ports_obj[try_arg]
+            except KeyError:
+                arg = None
+            return arg
+
+        self.association_list = dict()
+        for subset in self.subsets:
+
+            # Generate a list of new svc profiles per port spec
+            port_rules = list()
+            for port_spec in subset["ports"]:
+                dst_port = verify_args(port_spec, "port")
+                protocol = verify_args(port_spec, "protocol").lower()
+                if protocol and dst_port:
+                    port_rules.append(Rule(action="allow",
+                                           dst_ports=[dst_port],
+                                           protocol=protocol))
+
+            # Create Profile for the subset
+            # TODO: better profile names
+            profile_name = "svc_%s_%s_port_%s_protocol_%s" % (
```

Comment:
Yep - we should fix the naming up here.

It looks like there are potentially multiple port/protocol combinations that could be exposed in a port spec.

So, an appropriate name might be: `%ns_svc_%svc_%protocol1%port1_...%protocolN%portN"`

Which might look something like this in practice: `production_svc_frontend_tcp80_udp80`